core/src/main/java/com/amazon/opendistroforelasticsearch/sql/expression/datetime/DateTimeFunction.java

```diff
@@ -72,17 +100,202 @@ private FunctionResolver date() {
         impl(nullMissingHandling(DateTimeFunction::exprDate), DATE, TIMESTAMP));
   }
 
+  /**
+   * Specify a start date and add a temporal amount to the date.
+   * The return type depends on the date type and the interval unit. Detailed supported signatures:
+   * (DATE, DATETIME/TIMESTAMP, INTERVAL) -> DATETIME
+   * (DATE, LONG) -> DATE
+   * (DATETIME/TIMESTAMP, LONG) -> DATETIME
+   */
+  private FunctionResolver date_add() {
+    return define(BuiltinFunctionName.DATE_ADD.getName(),
+        impl(nullMissingHandling(DateTimeFunction::exprAddDateInterval), DATE, DATE, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprAddDateInterval), DATETIME, DATE, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprAddDateInterval),
+            DATETIME, DATETIME, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprAddDateInterval),
+            DATETIME, TIMESTAMP, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprAddDateDays), DATE, DATE, LONG),
+        impl(nullMissingHandling(DateTimeFunction::exprAddDateDays), DATETIME, DATETIME, LONG),
+        impl(nullMissingHandling(DateTimeFunction::exprAddDateDays), DATETIME, TIMESTAMP, LONG)
+    );
+  }
+
+  /**
+   * Specify a start date and subtract a temporal amount to the date.
+   * The return type depends on the date type and the interval unit. Detailed supported signatures:
+   * (DATE, DATETIME/TIMESTAMP, INTERVAL) -> DATETIME
+   * (DATE, LONG) -> DATE
+   * (DATETIME/TIMESTAMP, LONG) -> DATETIME
+   */
+  private FunctionResolver date_sub() {
+    return define(BuiltinFunctionName.DATE_SUB.getName(),
+        impl(nullMissingHandling(DateTimeFunction::exprSubDateInterval), DATE, DATE, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprSubDateInterval), DATETIME, DATE, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprSubDateInterval),
+            DATETIME, DATETIME, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprSubDateInterval),
+            DATETIME, TIMESTAMP, INTERVAL),
+        impl(nullMissingHandling(DateTimeFunction::exprSubDateDays), DATE, DATE, LONG),
+        impl(nullMissingHandling(DateTimeFunction::exprSubDateDays), DATETIME, DATETIME, LONG),
+        impl(nullMissingHandling(DateTimeFunction::exprSubDateDays), DATETIME, TIMESTAMP, LONG)
+    );
+  }
+
+  /**
+   * DAY(DATE). return the day of the month (1-31).
+   */
+  private FunctionResolver day() {
```

Comment:
e.g. day(date_add("2020-09-21", 1))
this will throw exception because day function can't accept datetime as paramater?

Reply:
added datetime & timestamp for all functions